lib/Language/Souffle/Experimental.hs

```diff
@@ -991,6 +1005,24 @@ accessorNames _ = case toStrings (Proxy :: Proxy (AccessorNames a)) of
 --   Only tuples containing up to 10 elements are currently supported.
 type Tuple ctx ts = TupleOf (MapType (Term ctx) ts)
 
+class GenVars (ts :: [Type]) where
+  genVars :: Proxy ts -> Tuple 'Relation ts
+
+varI :: Int -> Term 'Relation a
+varI i = VarTerm $ "x" <> T.pack (show i)
+
+instance GenVars '[t] where
+  genVars _ = varI 0
+
+instance GenVars '[t1, t2] where
+  genVars _ = (varI 0, varI 1)
+
```

Comment:
This sucks big-time. It should be an inductive definition, instead of requiring separate instances for each tuple length. Also, it should run in one of your variable number-tracking monads, but I decided that life's too short to worry about that for this illustration.

Reply:
Yes, tuples are really annoying to work with in Haskell :( main reason I used them is to keep the syntax the same as in Datalog. For now I limited it to 10 instances which is kinda "meh", but it gets the job done.

If you know a better way of handling raw tuples in combination with typeclasses, let me know!

Reply:
Well the straightforward way would be to use an `HList`-like product type indexed by its coordinates, through and through, but then you have to use `x :> y :> Nil` instead of `(x, y)`.

You could carry a list of types everywhere, and convert to a Haskell tuple at the last minute, but that becomes unwieldy on the way back I think -- because of one-tuples, you can't easily have a bijection between `Product ts` and `(t1, ..., tn)` :( And if you only want to do the conversion where you *do* know the `ts` ,that means you would need a special apply operator instead of ` `. If you are willing to do that, I think you can get both

    foo |- \x y -> ...

and 
 
    ... foo |$| (x, y) ...